### DISCUSSION TITLE
react-hook-form 참고 자료

### DISCUSSION THREAD

**Discussion by Lim-JiSeon** [Team-PalPalHae-Dealight/Team-PalPalHae-Dealight-FE#137] [General]
이 자료는 사용 방법에 초점을 맞춰 작성했습니다. 

```ts
import { useForm, SubmitHandler } from "react-hook-form"

type Inputs = {
  example: string
  exampleRequired: string
}

export default function App() {
  const {
    register,
    handleSubmit,
    watch,
    formState: { errors },
  } = useForm<Inputs>()
  const onSubmit: SubmitHandler<Inputs> = (data) => console.log(data)

  console.log(watch("example")) // watch input value by passing the name of it

  return (
    /* "handleSubmit" will validate your inputs before invoking "onSubmit" */
    <form onSubmit={handleSubmit(onSubmit)}>
      {/* register your input into the hook by invoking the "register" function */}
      <input defaultValue="test" {...register("example")} />

      {/* include validation with required or other standard HTML validation rules */}
      <input {...register("exampleRequired", { required: true })} />
      {/* errors will return when field validation fails  */}
      {errors.exampleRequired && <span>This field is required</span>}

      <input type="submit" />
    </form>
  )
}
```

## 1. 입력받을 값 type 지정하기 
```ts
type Inputs = {
  example: string
  exampleRequired: string
}
```

## 2. useForm에서 필요한 기능 불러오기
```ts
const {
    register,
    handleSubmit,
    watch,
    formState: { errors },
  } = useForm<Inputs>()
```
- register : 꼭 이름(name) 지정해서 등록해주세요. register로 할 수 있는 일들은 [참고 자료](https://react-hook-form.com/get-started#Registerfields) 참고해주세요. (기본적으로 yup을 이용해 타당성 검사를 한다면 등록 외에는 크게 기능을 사용하지 않을 듯 합니다.)
- handleSubmit : form을 제출할 때 실행할 기능을 handleSubmit에 넣어주세요. 예시에서는 onSubmit을 만들어 넘겼습니다. 예시를 참고해서 작성해주세요.
- watch : 이 함수를 통해 값에 접근할 수 있으며, 입력 값의 변화를 감지할 수 있습니다. (watch("example") 이런식으로 example에만 접근할 수도 있고, watch()를 써서 모든 데이터에 접근할 수도 있습니다.)
- formState: 보통 formData에서도 error를 많이 사용하며, 입력 validation을 체크할 때 error 처리해주시면 됩니다. 그 사용 방식은 예시 코드처럼 errors 객체의 해당 네임을 가진 값의 여부를 체크한 뒤 에러가 있을 경우 보여줄 에러 메시지를 참고해주시면 됩니다. 

## 3. 주의할 점
- react hook form은 기본적으로 defaultValue를 설정하지 않고 입력받지 않으면 undefined 입니다.
- 제가 이미 install 했기 때문에 pacakage에 react hook form이 있을 텐데 확인하시고, 없다면 dev pull and install 해주세요.
- 에러 조건 즉 validation은 yup으로 schema를 만들어 주세요. 만드는 방법은 formik에서 yup 부분을 참고해주세요.
- react hook form에서 yup을 사용하기 위해서는 resolver: yupResolver를 사용해야 합니다. 사용 방법은 아래를 참고해주세요. 

```ts
import { yupResolver } from '@hookform/resolvers/yup' // 이 부분은 설치해주세요. 설치하지 않았습니다.

const schema = 블라블라


const {
    register,
    handleSubmit,
    formState: { errors },
} = useForm<아까 정의한 input 타입>({ resolver: yupResolver(schema) });
```

참고자료
[react hook form 공식 문서](https://react-hook-form.com/get-started#Quickstart)
[react hook form yup 적용](https://medium.com/@msgold/creating-a-react-form-using-react-hook-form-and-yup-in-typescript-640168c5ed57)

**Comment by euan-lee** [comment]
자료 정리 감사합니다!
